File: frontend/src/core/hub.ts
/* 🌐 Hub client (spec §9.6) — exact shapes from weight_stream/server/hub.py
   (DownloadTask.to_dict / DownloadManager._parse_search) and the /v1/hub/*
   routes in api_server.py. Progress telemetry is ALWAYS server-computed
   (ADR-003): bytes/percent/speed_bps/eta_s are displayed as delivered —
   never smoothed, estimated client-side, or fabricated here. */
import { apiJSON } from './api'

export interface HubFile {
  filename: string
  /** quant parsed server-side from the filename; null when unparseable */
  quant: string | null
  /** parameter-size label (7B, 8X7B) parsed from repo+file name; null = unknown */
  size_label: string | null
}

export interface HubSearchResult {
  repo_id: string
  /** author (namespace) the server derived from repo_id; null for bare names */
  author: string | null
  downloads: number | null
  likes: number | null
  last_modified: string | null
  /** passed through from the (expanded) HF search response; null when absent */
  pipeline_tag: string | null
  /** passed through from the HF search response; [] when absent */
  tags: string[]
  gguf: true
  files: HubFile[]
}

/* ── On-demand model detail (GET /v1/hub/model/{repo}) ──────────────
   Every field mirrors weight_stream/server/hub.py::_build_detail. Fields HF
   does not provide are null — this client NEVER fills them in (ADR-003). */

export interface HubShard {
  /** 1-based position within the quant's shard set */
  index: number
  total: number
}

export interface HubDetailFile {
  filename: string
  /** real byte size from HF tree/main; null if HF omitted it */
  bytes: number | null
  quant: string | null
  size_label: string | null
  shard: HubShard | null
}

export interface HubNonGguf {
  filename: string
  bytes: number | null
  type: string
}

export interface HubQuantGroup {
  quant: string | null
  files: HubDetailFile[]
  /** sum of the shard bytes; null if any part's size is unknown */
  total_bytes: number | null
  sharded: boolean
  /** ordered per-shard byte sizes; null when not sharded / incomplete */
  per_shard_bytes: number[] | null
}

export interface HubModelDetail {
  repo_id: string
  author: string | null
  published_at: string | null
  updated_at: string | null
  downloads: number | null
  likes: number | null
  pipeline_tag: string | null
  tags: string[]
  library: string | null
  description: string | null
  base_model: string | string[] | null
  /** only ever a real value from cardData/tags; otherwise null */
  context_length: number | null
  files: HubDetailFile[]
  non_gguf: HubNonGguf[]
  quants: HubQuantGroup[]
}

/* ── Curated "proven on this rig" recommendations ───────────────────
   Served from weight_stream/server/recommended.py — every entry is backed
   by a measured experiment (research/experiments/) incl. the Thai quality
   gate, and carries the EXACT files that were measured (byte sizes verified
   against HF at curation time). Numbers are from THIS machine — honest
   caveat rendered alongside (ADR-003). */

export interface HubRecommendedFile {
  filename: string
  /** real byte size of this file (verified against HF at curation time) */
  bytes: number
}

/** Human-readable text that ships in both project languages; the UI picks
    the active one, the measured numbers stay a single source of truth. */
export interface HubLocaleText {
  en: string
  th: string
}

export interface HubRecommendedQuant {
  quant: string
  /** the exact download set — the Hub downloads precisely these */
  files: HubRecommendedFile[]
  total_bytes: number
  /** measured tok/s range on the reference rig; null = not measured */
  tok_s_min: number | null
  tok_s_max: number | null
  /** Thai quality gate: correct/total (9 fixed questions, EXP-009 set) */
  thai_correct: number | null
  thai_total: number | null
  /** Thai tonal discriminator: correct/total (the project's quality floor) */
  thai_tonal_correct: number | null
  thai_tonal_total: number | null
  /** repo-relative path to the evidence (research/experiments/EXP-*) */
  experiment: string
  /** free-form measurement/usage note, localised */
  notes: HubLocaleText | null
  /** llama-server flags the measurement used (e.g. MTP); null = none */
  flags: string | null
}

export type HubRecommendedRole = 'thai' | 'speed' | 'balanced'

export interface HubRecommendedEntry {
  repo_id: string
  name: string
  arch: string
  /** drives the badge colour/icon in the UI */
  role: HubRecommendedRole
  tagline: HubLocaleText
  quants: HubRecommendedQuant[]
}

export interface HubRecommendedResponse {
  recommended: HubRecommendedEntry[]
  count: number
}

/** Pick the active language from a locale dict; falls back to English, then
    to whatever exists (a missing locale never blanks a label). */
export function pickLocaleText(txt: HubLocaleText, lang: string): string {
  return txt[lang as keyof HubLocaleText] ?? txt.en ?? txt.th
}

export function hubRecommended(): Promise<HubRecommendedResponse> {
  return apiJSON<HubRecommendedResponse>('/v1/hub/recommended', undefined, { timeoutMs: 10_000 })
}

/* ── In-app experiment evidence (GET /v1/research/experiment/{path}) ──
   The Evidence button opens the experiment record INSIDE the app (the repo
   is private — no GitHub link needed). The server validates the path stays
   under research/experiments/ and returns the folder's markdown files. */

export interface ResearchExperimentFile {
  /** e.g. setup.md / results.md / analysis.md */
  name: string
  markdown: string
}

export interface ResearchExperiment {
  /** the validated experiment dir (research/experiments/EXP-XXX-…/) */
  path: string
  /** real byte sizes per file; 0/null only when the read failed */
  files: ResearchExperimentFile[]
}

export function researchExperiment(expPath: string): Promise<ResearchExperiment> {
  const enc = expPath.split('/').map(encodeURIComponent).join('/')
  return apiJSON<ResearchExperiment>(`/v1/research/experiment/${enc}`, undefined, { timeoutMs: 10_000 })
}

export function hubModel(repoId: string): Promise<HubModelDetail> {
  // repo_id is server-sourced → encode each path segment, never interpolate raw
  const enc = repoId.split('/').map(encodeURIComponent).join('/')
  return apiJSON<HubModelDetail>(`/v1/hub/model/${enc}`, undefined, { timeoutMs: 20_000 })
}

export interface HubSearchResponse {
  results: HubSearchResult[]
  count: number
  /** HF cursor for the next page; null when unavailable / last page. */
  next_cursor: string | null
}

export type HubTaskStatus = 'queued' | 'downloading' | 'done' | 'failed' | 'cancelled'

export interface HubTask {
  id: string
  repo_id: string
  filename: string
  target_dir: string
  target_path: string
  status: HubTaskStatus
  bytes_downloaded: number
  total_bytes: number | null
  /** REAL byte size of the file on disk for a completed download; null
      otherwise (stat'ed server-side at serialization time — honest). */
  file_size: number | null
  /** null until total_bytes is known — render "–", never guess */
  percent: number | null
  speed_bps: number | null
  eta_s: number | null
  error: string | null
  created_at: number
  updated_at: number
}

export const HUB_TERMINAL: readonly HubTaskStatus[] = ['done', 'failed', 'cancelled']

export type HubSort = 'downloads' | 'likes' | 'recent'

export function hubSearch(q: string, sort: HubSort, limit = 20): Promise<HubSearchResponse> {
  const qs = new URLSearchParams({ q, sort, limit: String(limit) })
  // HF can be slow; the server itself times out at 10s.
  return apiJSON<HubSearchResponse>(`/v1/hub/search?${qs}`, undefined, { timeoutMs: 20_000 })
}

/** Paged search used by the Hub "Latest" feed: asks the server for real HF
    cursor pagination (paginate=1) and returns the ``next_cursor`` too. */
export function hubSearchPage(
  sort: HubSort,
  cursor: string | null,
  limit = 20,
): Promise<HubSearchResponse> {
  const qs = new URLSearchParams({
    sort,
    limit: String(limit),
    paginate: '1',
  })
  if (cursor) qs.set('cursor', cursor)
  return apiJSON<HubSearchResponse>(`/v1/hub/search?${qs}`, undefined, { timeoutMs: 20_000 })
}

export function hubDownload(repoId: string, filename: string, targetDir?: string): Promise<HubTask> {
  return apiJSON<HubTask>(
    '/v1/hub/download',
    { method: 'POST', body: JSON.stringify({ repo_id: repoId, filename, target_dir: targetDir ?? null }) },
    { timeoutMs: 30_000 },
  )
}

export function hubDownloads(): Promise<{ downloads: HubTask[]; count: number }> {
  return apiJSON('/v1/hub/downloads', undefined, { timeoutMs: 10_000 })
}

export function hubCancel(taskId: string): Promise<HubTask> {
  return apiJSON<HubTask>(`/v1/hub/download/${encodeURIComponent(taskId)}/cancel`, { method: 'POST' }, { timeoutMs: 10_000 })
}

/** v1.1 resume: re-queues a cancelled/failed task; the server appends the
    remaining bytes to the kept ``.part`` via HTTP Range (never byte 0). */
export function hubResume(taskId: string): Promise<HubTask> {
  return apiJSON<HubTask>(`/v1/hub/download/${encodeURIComponent(taskId)}/resume`, { method: 'POST' }, { timeoutMs: 10_000 })
}

/** v1.1 delete: removes a task (stops a running worker, drops its .part).
    With `deleteFile` the server ALSO removes the completed .gguf from disk
    (only for done tasks whose model is not loaded); `file_deleted` in the
    response says honestly whether the file was removed. */

/** Cross-feature model references (the server scans what it owns —
    assistants; conversations live client-side so the UI counts those
    itself). Used to warn before a model file is deleted. */
export interface HubReferencedBy {
  /** assistant NAMES pinned to this download's suggested model id */
  assistants: string[]
}

export interface HubDeleteResult {
  status: string
  id: string
  file_deleted: boolean
  /** scanned live at delete time (catches cross-tab assistant edits) */
  referenced_by: HubReferencedBy
}

export function hubDelete(taskId: string, deleteFile = false): Promise<HubDeleteResult> {
  return apiJSON(
    `/v1/hub/download/${encodeURIComponent(taskId)}/delete`,
    deleteFile ? { method: 'POST', body: JSON.stringify({ delete_file: true }) } : { method: 'POST' },
    { timeoutMs: 10_000 },
  )
}

/** v1.1 clear-finished: remove every terminal task (done/failed/cancelled)
    in one call; active downloads are kept. With `deleteFile` the server ALSO
    deletes the model files of completed downloads, except those of currently
    loaded models (reported in `files_skipped`). */
export interface HubClearResult {
  status: string
  removed: string[]
  files_deleted: string[]
  files_skipped: string[]
  /** task id → references; present only when deleteFile was requested
      (only then are model files at risk). */
  referenced_by?: Record<string, HubReferencedBy>
}

export function hubClear(deleteFile = false): Promise<HubClearResult> {
  return apiJSON(
    '/v1/hub/downloads/clear',
    deleteFile ? { method: 'POST', body: JSON.stringify({ delete_file: true }) } : { method: 'POST' },
    { timeoutMs: 10_000 },
  )
}

/** v1.1 reveal: ask the server (same machine as the browser) to open the OS
    file manager showing a completed download's file. 404 unknown / 409 not
    finished / 403 outside the allowed model dirs / 500 launcher failed. */
export function hubReveal(taskId: string): Promise<{ status: string; path: string }> {
  return apiJSON(
    `/v1/hub/download/${encodeURIComponent(taskId)}/reveal`,
    { method: 'POST' },
    { timeoutMs: 10_000 },
  )
}

/** Visibility-aware SSE subscription to a download's REAL progress.

    Uses the native EventSource (no library — bundle budget). The server
    yields the current task state immediately on connect, then every 0.5s
    until a terminal status, so a (re)connect after the tab was hidden
    catches up at once. Returns `stop()` — call it on unmount AND whenever
    the tab becomes hidden (the brief requires pause-when-hidden). */
export function subscribeHubProgress(
  taskId: string,
  onFrame: (task: HubTask) => void,
  onError?: () => void,
): () => void {
  let es: EventSource | null = null
  let closed = false

  const open = () => {
    if (closed || document.hidden) return
    es = new EventSource(`/v1/hub/progress/${encodeURIComponent(taskId)}`)
    es.onmessage = (ev) => {
      try {
        onFrame(JSON.parse(ev.data) as HubTask)
      } catch {
        /* a malformed frame is not fatal — the next tick carries full state */
      }
    }
    es.onerror = () => {
      // The server CLOSES the stream after the terminal frame, which also
      // fires onerror (readyState CLOSED) — that is the happy path. Only
      // surface errors while the stream should still be live.
      es?.close()
      es = null
      if (!closed && !document.hidden) onError?.()
    }
  }

  const onVisibility = () => {
    if (document.hidden) {
      es?.close()
      es = null
    } else if (!closed) {
      open() // resume: server replays the current state immediately
    }
  }
  document.addEventListener('visibilitychange', onVisibility)

  open()
  return () => {
    closed = true
    es?.close()
    es = null
    document.removeEventListener('visibilitychange', onVisibility)
  }
}

/* ── display helpers (pure formatting, no fabricated data) ── */

export function fmtBytes(n: number | null | undefined): string {
  if (n == null) return '–'
  if (n < 1024) return `${n} B`
  const units = ['KB', 'MB', 'GB', 'TB']
  let v = n / 1024
  let i = 0
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024
    i++
  }
  return `${v >= 100 ? Math.round(v) : v.toFixed(1)} ${units[i]}`
}

export function fmtSpeed(bps: number | null | undefined): string | null {
  if (bps == null || !Number.isFinite(bps) || bps < 0) return null // honest: unknown
  return `${fmtBytes(bps)}/s`
}

export function fmtEta(seconds: number | null | undefined): string | null {
  if (seconds == null || !Number.isFinite(seconds) || seconds < 0) return null // honest: unknown
  const s = Math.round(seconds)
  if (s < 60) return `${s}s`
  const m = Math.floor(s / 60)
  if (m < 60) return `${m}m ${s % 60}s`
  return `${Math.floor(m / 60)}h ${m % 60}m`
}

/** repo "author/name" → author; repos without a slash have no separate author. */
export function repoAuthor(repoId: string): string | null {
  const i = repoId.indexOf('/')
  return i > 0 ? repoId.slice(0, i) : null
}

export function repoName(repoId: string): string {
  const i = repoId.indexOf('/')
  return i >= 0 ? repoId.slice(i + 1) : repoId
}

/** Safe external link to a repo on huggingface.co (server-sourced id →
    each path segment is URI-encoded; never interpolated raw). */
export function hfRepoUrl(repoId: string): string {
  return `https://huggingface.co/${repoId.split('/').map(encodeURIComponent).join('/')}`
}

/* ── Category + feature derivation (pure; from REAL HF pipeline_tag/tags) ──
   Honest: when HF gives no usable signal we return the neutral "other"
   category and zero feature badges — we never invent a category. */

export type HubCategoryId =
  | 'embedding'
  | 'vision'
  | 'audio'
  | 'code'
  | 'moe'
  | 'chat'
  | 'other'

export interface HubCategory {
  id: HubCategoryId
  emoji: string
  labelKey: string
}

/** Unquantized weight families — real, but usually far too large to want. */
const UNQUANTIZED = new Set(['F16', 'BF16', 'F32'])
export function isUnquantized(quant: string | null): boolean {
  return quant != null && UNQUANTIZED.has(quant.toUpperCase())
}

export function modelCategory(
  pipelineTag: string | null | undefined,
  tags: string[] | null | undefined,
): HubCategory {
  const pt = (pipelineTag || '').toLowerCase()
  const t = (tags || []).map((x) => String(x).toLowerCase())
  const has = (...needles: string[]) => t.some((tag) => needles.some((n) => tag.includes(n)))
  if (pt.includes('embedding') || pt.includes('feature-extraction') || pt.includes('sentence-similarity') || has('embedding', 'embeddings')) {
    return { id: 'embedding', emoji: '🔢', labelKey: 'hub.catEmbedding' }
  }
  if (pt.includes('image') || pt.includes('vision') || has('vision', 'image-to-text', 'multimodal')) {
    return { id: 'vision', emoji: '👁️', labelKey: 'hub.catVision' }
  }
  if (pt.includes('audio') || pt.includes('speech') || has('audio', 'tts', 'asr', 'speech')) {
    return { id: 'audio', emoji: '🎧', labelKey: 'hub.catAudio' }
  }
  if (pt.includes('code') || has('code')) {
    return { id: 'code', emoji: '💻', labelKey: 'hub.catCode' }
  }
  if (has('moe', 'mixture-of-experts')) {
    return { id: 'moe', emoji: '🧩', labelKey: 'hub.catMoe' }
  }
  if (pt.includes('text-generation') || pt.includes('conversational') || has('chat', 'conversational')) {
    return { id: 'chat', emoji: '💬', labelKey: 'hub.catChat' }
  }
  return { id: 'other', emoji: '🤖', labelKey: 'hub.catOther' }
}

/** Relevant, decision-useful feature hints derived from tags (max 4). */
export function modelFeatures(tags: string[] | null | undefined): string[] {
  const t = (tags || []).map((x) => String(x).toLowerCase())
  const out: string[] = []
  if (t.some((x) => x.includes('function-calling') || x.includes('function_calling') || x.includes('tool'))) {
    out.push('hub.featTools')
  }
  if (t.some((x) => x.includes('vision') || x.includes('image'))) out.push('hub.featVision')
  if (t.some((x) => x === 'th' || x.includes('thai'))) out.push('hub.featThai')
  if (t.some((x) => x.includes('moe') || x.includes('mixture-of-experts'))) out.push('hub.featMoe')
  return out.slice(0, 4)
}

File: weight_stream/server/routes/hub.py
"""Hub routes: Hugging Face search/recommended, downloads, research."""
from __future__ import annotations

import asyncio
import json
import os
from typing import Optional

from fastapi import APIRouter, FastAPI, HTTPException
from fastapi.responses import StreamingResponse

from ..config import get_model_search_dirs
from ..hub import DownloadManager, HubUpstreamError, HubValidationError
from ..recommended import to_payload as recommended_payload
from ..research import ResearchValidationError, experiment as read_experiment
from ..schemas import (
    HubDownloadRequest,
    HubDeleteRequest,
    HubClearRequest,
)
from .context import ServerContext


def _assistant_refs_batch(filenames: list[str]) -> dict[str, list[str]]:
    """filename → assistant names pinned to that file's suggested model id.

    Model references across features are keyed by the SUGGESTED model id
    (basename without ``.gguf`` — the same rule as the Console's
    ``suggestModelId`` used when loading a downloaded model and when a
    conversation is created against it). So an assistant pinned to that id
    silently loses its model if the file is deleted — the delete dialogs
    warn about exactly these references.

    ONE store read for the whole batch (``clear`` can carry several done
    tasks — per-task scans would re-read every assistant JSON N times). The
    store is read LIVE at request time, so an assistant created in another
    tab counts. Returns {} on any store problem — a reference-scan failure
    must never block a delete.
    """
    wanted: dict[str, str] = {}
    for f in filenames:
        model_id = os.path.splitext(os.path.basename(f))[0]
        if model_id:
            wanted[f] = model_id
    if not wanted:
        return {}
    try:
        from ..assistants import get_assistant_store
        store_list = get_assistant_store().list()
    except Exception:
        return {}
    by_id: dict[str, list[str]] = {m: [] for m in set(wanted.values())}
    for a in store_list:
        mid = a.get("model_id") or ""
        if mid in by_id:
            by_id[mid].append(a.get("name") or a.get("id", "?"))
    return {f: by_id[m] for f, m in wanted.items()}


def _assistants_referencing(filename: str) -> list[str]:
    """Assistant names for ONE file (see ``_assistant_refs_batch``)."""
    return _assistant_refs_batch([filename]).get(filename, [])


def _reveal_in_explorer(path: str) -> dict:
    """Reveal a file in the OS file manager (server-side subprocess).

    Opens the parent folder with the file selected where the platform
    supports it (Windows ``explorer /select``, macOS ``open -R``); Linux
    falls back to opening the folder. Returns ``{"error": ...}`` honestly
    when the shell command fails — never a fake success.

    Known platform quirk (documented, not worked around): ``explorer``
    mis-parses ``/select,<path>`` when the path itself contains a comma
    (GGUF model paths almost never do; the list-form Popen keeps the arg
    intact with no shell involved, so only the comma case misbehaves).
    """
    import subprocess
    import sys
    folder = os.path.dirname(path)
    try:
        if sys.platform == "win32":
            # explorer /select needs the comma syntax; quoting is handled
            # by Popen's list form (no shell involved).
            subprocess.Popen(["explorer", f"/select,{os.path.normpath(path)}"])
        elif sys.platform == "darwin":
            subprocess.Popen(["open", "-R", path])
        else:
            subprocess.Popen(["xdg-open", folder])
        return {"ok": True}
    except Exception as e:
        return {"error": str(e)}


def build_router(app: FastAPI, ctx: ServerContext) -> APIRouter:
    """Register hub + research routes."""
    router = APIRouter()
    hub_manager: Optional[DownloadManager] = ctx.hub_manager

    def _hub() -> DownloadManager:
        if hub_manager is None:
            raise HTTPException(status_code=500, detail="hub manager not configured")
        return hub_manager

    @router.get("/v1/hub/search")
    async def hub_search(
        q: str = "",
        sort: str = "downloads",
        limit: int = 20,
        cursor: str | None = None,
        paginate: int = 0,
    ):
        """
        Search Hugging Face for GGUF models (filtered to GGUF only), with
        quant + parameter-size parsed from each file's name. Results are
        cached in-memory for 5 minutes. HF unreachable → 502 (never a fake
        list). `sort` ∈ downloads|likes|recent.

        Optional cursor pagination for the Hub "Latest" feed: pass
        `paginate=1` (and optionally `cursor` for a page after the first) to
        also receive `next_cursor` in the response, threaded through the real
        HF `Link: rel="next"` header. The plain single-page path is unchanged.
        """
        hm = _hub()
        try:
            if paginate:
                page = hm.search_with_cursor(q=q, sort=sort, limit=limit, cursor=cursor)
                results = page["results"]
                return {
                    "results": results,
                    "count": len(results),
                    "next_cursor": page["next_cursor"],
                }
            results = hm.search(q=q, sort=sort, limit=limit)
            return {"results": results, "count": len(results), "next_cursor": None}
        except HubUpstreamError as e:
            raise HTTPException(status_code=502, detail=str(e))

    @router.get("/v1/hub/recommended")
    async def hub_recommended():
        """Curated models proven on this reference rig (server/recommended.py).

        Static data — no network call. Every entry is backed by a measured
        experiment (``research/experiments/``) with the Thai quality gate, and
        carries the EXACT download files that were measured so users fetch the
        right quant. See the module docstring for the honest caveat (numbers
        are from this machine; other hardware will differ).
        """
        return recommended_payload()

    @router.get("/v1/research/experiment/{exp_path:path}")
    async def research_experiment(exp_path: str):
        """Serve one experiment's markdown record for the in-app Evidence
        viewer (research/experiments/). Path validated by containment — no
        traversal, only ``*.md`` files ever read (server/research.py).
        """
        try:
            return read_experiment(exp_path)
        except ResearchValidationError as e:
            raise HTTPException(status_code=e.status, detail=str(e))

    @router.get("/v1/hub/model/{repo_id:path}")
    async def hub_model(repo_id: str):
        """
        On-demand model detail (P5.1): aggregate HF model metadata + per-file
        byte sizes + shard/quant grouping for one repo. Cached ~15 min. HF
        unreachable → 502 (never a fake/empty 200). Fields HF omits are null.
        """
        try:
            return _hub().model_detail(repo_id)
        except HubValidationError as e:
            raise HTTPException(status_code=e.status, detail=str(e))
        except HubUpstreamError as e:
            raise HTTPException(status_code=502, detail=str(e))

    @router.post("/v1/hub/download", status_code=202)
    async def hub_download(body: HubDownloadRequest):
        """
        Start a background GGUF download and return the task (poll
        `/v1/hub/progress/{id}` or `/v1/hub/downloads`).

        Security: `target_dir` must resolve inside an allowed model dir
        (traversal / absolute-outside / symlink-escape → 403); `filename`
        must be a plain `*.gguf` name (else 400). Writes are atomic
        (`.part` → rename) and size-guarded. No auth in v1 — isolate the
        server (see API Docs note, P5).
        """
        hm = _hub()
        try:
            task = hm.create_download(body.repo_id, body.filename, body.target_dir)
        except HubValidationError as e:
            raise HTTPException(status_code=e.status, detail=str(e))
        hm.schedule_download(task)
        return task.to_dict()

    @router.get("/v1/hub/downloads")
    async def hub_downloads():
        """List all download tasks with their latest status/progress."""
        items = _hub().list_tasks()
        return {"downloads": items, "count": len(items)}

    @router.get("/v1/hub/progress/{task_id}")
    async def hub_progress(task_id: str):
        """
        SSE stream of a download's REAL progress (bytes/percent/speed_bps/
        eta_s/status) until it reaches a terminal state (done/failed/cancelled).
        """
        task = _hub().get_task(task_id)
        if task is None:
            raise HTTPException(status_code=404, detail=f"download task {task_id} not found")

        async def event_generator():
            while True:
                yield f"data: {json.dumps(task.to_dict(), ensure_ascii=False)}\n\n"
                if task.status in ("done", "failed", "cancelled"):
                    break
                await asyncio.sleep(0.5)

        return StreamingResponse(
            event_generator(),
            media_type="text/event-stream",
            headers={
                "X-Accel-Buffering": "no",
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
            },
        )

    @router.post("/v1/hub/download/{task_id}/cancel")
    async def hub_cancel(task_id: str):
        """
        Cancel a download. Sets the cancel flag; the worker stops within one
        chunk and the partial ``.part`` is KEPT so the task can be resumed.
        Idempotent for already-terminal tasks.
        """
        task = _hub().cancel(task_id)
        if task is None:
            raise HTTPException(status_code=404, detail=f"download task {task_id} not found")
        return task.to_dict()

    @router.post("/v1/hub/download/{task_id}/resume")
    async def hub_resume(task_id: str):
        """
        Resume a cancelled/failed download (v1.1): re-queues the task; the
        worker appends the remaining bytes to the kept ``.part`` via HTTP
        ``Range`` instead of re-downloading from byte 0. 404 unknown task;
        409 when the task is not resumable (active or done).
        """
        hm = _hub()
        try:
            task = hm.resume(task_id)
        except ValueError as e:
            raise HTTPException(status_code=409, detail=str(e))
        if task is None:
            raise HTTPException(status_code=404, detail=f"download task {task_id} not found")
        hm.schedule_download(task)
        return task.to_dict()

    @router.post("/v1/hub/download/{task_id}/delete")
    async def hub_delete(task_id: str, body: Optional[HubDeleteRequest] = None):
        """
        Delete a download task from the manager (v1.1): stops a running
        worker and removes the partial ``.part``. By default the final
        ``.gguf`` of a completed task is left on disk; pass
        ``{"delete_file": true}`` to ALSO delete the model file (only for
        ``done`` tasks whose file is inside an allowed model dir and whose
        model is not currently loaded). Returns ``file_deleted`` honestly.
        """
        hm = _hub()
        delete_file = bool(body and body.delete_file)
        task = hm.get_task(task_id)
        if task is None:
            raise HTTPException(status_code=404, detail=f"download task {task_id} not found")
        if delete_file:
            if task.status != "done":
                raise HTTPException(
                    status_code=409,
                    detail="only a completed download has a model file to delete",
                )
            # never delete a model that a backend is holding open — removing
            # it would break the running session (checked BEFORE the delete).
            # Compare NORMALIZED paths (realpath + case-fold) so a model
            # loaded through a different spelling/symlink still matches.
            target_real = os.path.normcase(os.path.realpath(task.target_path))
            loaded = await ctx.manager.list_models()
            if any(
                os.path.normcase(os.path.realpath(getattr(m, "path", "") or "")) == target_real
                for m in loaded
            ):
                raise HTTPException(
                    status_code=409,
                    detail="this model is currently loaded — unload it before deleting the file",
                )
        hm.delete(task_id, delete_file=delete_file)
        file_deleted = delete_file and not os.path.exists(task.target_path)
        # which features reference this model's suggested id (conversations
        # live client-side, so the server reports assistants only — the UI
        # counts conversations itself). Honest at delete time: a reference
        # created in another tab after the dialog opened is still reported.
        # Only scanned when a file is actually at risk (delete_file) — the
        # keep-file path never reads the assistant store.
        refs = _assistants_referencing(task.filename) if delete_file else []
        return {
            "status": "deleted",
            "id": task.id,
            "file_deleted": file_deleted,
            "referenced_by": {"assistants": refs},
        }

    @router.post("/v1/hub/downloads/clear")
    async def hub_clear(body: Optional[HubClearRequest] = None):
        """
        Remove every FINISHED download (done/failed/cancelled) at once
        (v1.1): the panel's "clear finished" action. Active downloads are
        kept. Pass ``{"delete_file": true}`` to ALSO delete the model files
        of completed downloads — except those of currently loaded models,
        which are skipped and reported in ``files_skipped`` (never removed
        under a running backend). Returns the honest summary.
        """
        hm = _hub()
        delete_file = bool(body and body.delete_file)
        protected: set = set()
        if delete_file:
            # same normalized-path rule as the single-task delete endpoint
            loaded = await ctx.manager.list_models()
            protected = {
                os.path.normcase(os.path.realpath(getattr(m, "path", "") or ""))
                for m in loaded
            }
        # snapshot the DONE tasks' filenames BEFORE the clear so the response
        # can map task id → assistant references (the tasks are popped inside
        # clear(); their filenames would otherwise be gone).
        done_files = {
            t["id"]: t["filename"]
            for t in hm.list_tasks()
            if t.get("status") == "done"
        }
        result = hm.clear(delete_file=delete_file, protected_paths=protected)
        if delete_file and done_files:
            # ONE store read for the whole batch, then task_id → references
            # for every done task the clear removed (conversations live
            # client-side, so only assistants are reported). A task that
            # finished between the snapshot and the clear simply has no
            # entry — advisory only, never blocks.
            refs_by_file = _assistant_refs_batch(list(done_files.values()))
            result["referenced_by"] = {
                tid: {"assistants": refs_by_file.get(fname, [])}
                for tid, fname in done_files.items()
                if tid in result.get("removed", [])
            }
        return result

    @router.post("/v1/hub/download/{task_id}/reveal")
    async def hub_reveal(task_id: str):
        """
        Open the OS file manager showing a COMPLETED download's file (v1.1).

        The server and the browser run on the same machine, so this launches
        Explorer/Finder via a subprocess (Windows ``/select`` highlights the
        file; macOS ``open -R``; Linux opens the folder). Security: only
        tasks that finished with their file on disk, and the file must
        realpath-resolve inside an allowed model dir (same containment rule
        as delete) — revealing an arbitrary path is refused.
        """
        task = _hub().get_task(task_id)
        if task is None:
            raise HTTPException(status_code=404, detail=f"download task {task_id} not found")
        if task.status != "done" or not os.path.isfile(task.target_path):
            raise HTTPException(
                status_code=409,
                detail="only a completed download whose file is still on disk can be revealed",
            )
        real = os.path.realpath(task.target_path)
        allowed = [os.path.realpath(d) for d in get_model_search_dirs()]
        if not any(real == ra or real.startswith(ra + os.sep) for ra in allowed):
            raise HTTPException(
                status_code=403,
                detail="refusing to reveal a file outside the allowed model directories",
            )
        res = _reveal_in_explorer(real)
        if res.get("error"):
            raise HTTPException(status_code=500, detail=res["error"])
        return {"status": "revealed", "path": real}

    return router
